multitenancy-spring-boot-autoconfigure/src/main/java/org/axonframework/extensions/multitenancy/autoconfig/MultiTenantEventScheduler.java

```diff
@@ -0,0 +1,268 @@
+ *
+ * @author Stefan Dragisic
+ * @since 4.9.0
+ */
+public class MultiTenantEventScheduler implements EventScheduler, MultiTenantAwareComponent {
+
+    private final TenantEventSchedulerSegmentFactory tenantSegmentFactory;
+    private final TargetTenantResolver<EventMessage<?>> targetTenantResolver;
+
+    private final Map<TenantDescriptor, EventScheduler> tenantSegments = new ConcurrentHashMap<>();
+
+    /**
+     * Instantiate a {@link MultiTenantEventScheduler} based on the fields contained in the {@link Builder}.
+     * @param builder the {@link Builder} used to instantiate a {@link MultiTenantEventScheduler} instance
+     */
+    public MultiTenantEventScheduler(MultiTenantEventScheduler.Builder builder) {
+        builder.validate();
+        this.tenantSegmentFactory = builder.tenantSegmentFactory;
+        this.targetTenantResolver = builder.targetTenantResolver;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link MultiTenantEventScheduler}.
+     * @return a Builder to be able to create a {@link MultiTenantEventScheduler}
+     */
+    public static MultiTenantEventScheduler.Builder builder() {
+        return new MultiTenantEventScheduler.Builder();
+    }
+
+    /**
+     * Schedule the given {@code event} for publication at the given {@code triggerDateTime}.
+     * Message must contain MetaData with {@link TenantDescriptor} to resolve the tenant.
+     * Therefor only EventMessage type is supported.
+     *
+     * @param instant The moment to trigger publication of the event
+     * @param o           The event to publish
+     * @return the token to use when cancelling the schedule
+     */
+    @Override
+    public ScheduleToken schedule(Instant instant, Object o) {
+        return resolveTenant(o).schedule(instant, o);
+    }
+
+    /**
+     * Schedule the given {@code event} for publication after the given {@code duration}.
+     * Message must contain MetaData with {@link TenantDescriptor} to resolve the tenant.
+     * Therefor only EventMessage type is supported.
+     *
+     * @param duration The amount of time to wait before publishing the event
+     * @param o           The event to publish
+     * @return the token to use when cancelling the schedule
+     */
+    @Override
+    public ScheduleToken schedule(Duration duration, Object o) {
+        return resolveTenant(o).schedule(duration, o);
+    }
+
+    /**
+     * Cancel the publication of a scheduled event. This method is not supported for multi-tenant event scheduler.
+     * Get the tenant specific event scheduler and cancel the schedule there.
+     * See {@link #forTenant(TenantDescriptor)}.
+     *
+     * @param scheduleToken the token returned when the event was scheduled
+     */
+    @Override
+    public void cancelSchedule(ScheduleToken scheduleToken) {
+        throw new UnsupportedOperationException("Cancel schedule is not supported for multi-tenant event scheduler directly."
```

Comment:
Why not use the `TenantWrappedTransactionManager#getCurrentTenant` here too? Chances are high users will cancel schedules from within other message/event handlers I think.

Reply:
you are right Steven!